src/main/scala/firrtl/ir/Serializer.scala

```diff
@@ -0,0 +1,207 @@
+// See LICENSE for license details.
+
+package firrtl.ir
+
+import firrtl.constraint.Constraint
+
+object Serializer {
+  val NewLine = '\n'
+  val Indent = "  "
+
+  /** Converts a `FirrtlNode` into its string representation. */
+  def serialize(node: FirrtlNode): String = {
+    val builder = new StringBuilder()
+    val indent = 0
+    s(node)(builder, indent)
+    builder.toString()
+  }
+
+  /** Converts a `Constraint` into its string representation. */
+  def serialize(con: Constraint): String = {
+    val builder = new StringBuilder()
+    s(con)(builder)
+    builder.toString()
+  }
+
+  private def flattenInfo(infos: Seq[Info]): Seq[FileInfo] = infos.flatMap {
+    case NoInfo => Seq()
+    case f : FileInfo => Seq(f)
+    case MultiInfo(infos) => flattenInfo(infos)
+  }
+
+  //scalastyle:off cyclomatic.complexity method.length
+  private def s(node: FirrtlNode)(implicit b: StringBuilder, indent: Int): Unit = node match {
+    case NoInfo => // empty string
+    case FileInfo(info) => b ++= " @[" ; s(info) ; b ++= "]"
+    case MultiInfo(infos) =>
+      val ii = flattenInfo(infos)
+      b ++= " @[" ; s(ii, ", ") ; b ++= "]"
+    case s : StringLit => b ++= s.serialize
+
```

Comment:
Rather than a single function that takes `FirrtlNode`, I think it would be more clear and slightly faster to have 1 function per subtype, eg. `Statement`, `Expression`, `Type`, etc.

Reply:
I can change that!

Reply:
I implemented your requested change. It made the code a little bigger (~60 additional lines because I cannot use a polymorphic function for argument serialization anymore and thus had to add `sField` and `sInfo`) but it does improve things for the microbenchmark.

medium@JVM1.8: 850ms ± 10ms -> 802ms ± 5ms
medium@JVM14: 818ms ± 28ms -> 720ms ± 13ms